### ISSUE TITLE
[BUG] Model's performance discrepancy between evaluation and inference modes

### ISSUE THREAD

**Issue by sararb** [NVIDIA-Merlin/Transformers4Rec#525]
### Bug description

This BUG is a summary of the thread discussion happening in #506.  The main issue is the discrepancy of scores when applying the model using the masking to replace the last interaction with [MASK] embeddings and applying the same model without using the masking replacement. Instead of masking, the last interaction of the input sequence is explicitly set to `0` (to mimic the real-life inference scenario)

Given the input sequence, a subset of positions is replaced by special embeddings [MASK], and the model is trained to recover the original item-id from the corrupted input sequence.   During inference, we don't rely on data corruption (using [MASK]). Instead, we apply the model to the original test input sequence. Hence, we observe a discrepancy in scores because the model learned latent information from [MASKING] embeddings which is not used in the original test data. 

### Steps/Code to reproduce bug

1. I summarized the code shared in thread #506 in this [gist](https://gist.github.com/sararb/7e61b286aa612a2d707a59832bd948ff#file-replicate_t4rec_eval_metrics-py).
2. The code can be added to the notebook `02-End-to-end-session-based-with-Yoochoose-PyT` after the cell 10 to get the different test scores.

### Expected behavior
- The performance scores should match between the training/evaluation mode where data is corrupted with the [MASKING] embeddings and the test/inference mode where the masking is not applied to the original data. 

### Additional context
Some possible solutions to fix the discrepancy are: 
- Extend the test data with a dummy last interaction and use the model with `ignore_masking=False` to replace the last position with the [MASK] embeddings. 
- Replace the trainable masked embeddings tensor by the embeddings of the padding-idx `0`

**Comment by zuoxingdong** [comment]
@sararb 

As a quick workaround (not a fix though), the "consistency" can be ensured by setting masking embedding as zero embedding and turning it off for training.

```python
model.heads[0].body[0].masking.masked_item_embedding.requires_grad = False
model.heads[0].body[0].masking.masked_item_embedding.zero_()
```

I compared the effects with my own dataset:

```python
# w/ zero mask embedding
# test_/next-item/ndcg_at_10': 0.25335368514060974, 'test_/next-item/ndcg_at_20': 0.3103814423084259, 'test_/next-item/recall_at_10': 0.611253559589386, 'test_/next-item/recall_at_20': 0.8374492526054382
# eval_/next-item/ndcg_at_10': 0.4539324939250946, 'eval_/next-item/ndcg_at_20': 0.5021244287490845, 'eval_/next-item/recall_at_10': 0.7314428687095642, 'eval_/next-item/recall_at_20': 0.9207673072814941
# w/o zero mask embedding
# test_/next-item/ndcg_at_10': 0.17744575440883636, 'test_/next-item/ndcg_at_20': 0.268817275762558, 'test_/next-item/recall_at_10': 0.4537668228149414, 'test_/next-item/recall_at_20': 0.819256067276001
# eval_/next-item/ndcg_at_10': 0.456199049949646, 'eval_/next-item/ndcg_at_20': 0.5016158819198608, 'eval_/next-item/recall_at_10': 0.7311648726463318, 'eval_/next-item/recall_at_20': 0.9107589721679688
```
- `trainer.predict` (i.e. ignore_masking=False) -> `test_*` 
- `trainer.evaluate` (i.e. ignore_masking=True) -> `eval_*` 

The evaluations have similar performance, but the one with `zero mask embedding` has a better performance for `inference mode`

**Comment by sararb** [comment]
Thank you @zuoxingdong, for the quick workaround! To find a permanent fix for the inference mode, I am running a comparative benchmark with different setups, and I will include your proposed solution too. 

If I understand your setting correctly, you are using the same evaluation data for trainer.predict and trainer.evaluate, but trainer.predict has an additional processing step to mask the last interaction in each input session, right?

**Comment by zuoxingdong** [comment]
Hi @sararb , thank you so much for your great efforts in solving this issue!

Yes, the setting is exactly how you described above.
I'll check out the linked PR for your solution!
